### PR TITLE
ISSUE 1590 - Migration: Viewer, Model and rest of services

### LINKED ISSUE #1590: Migration: Viewer & Model

### PR DESCRIPTION
**Pull request by simplypixi**
This fixes #1590 , #1292, #1522 , #1589 

#### Changes
- Migration of Model & Viewer migration
We create two components: ViewerCanvas and ViewerGui.
- Migration of top level of routing
- Removing the rest of old Angular code
- tsconfig update
- tslint update
- libraries upgrade
Now the app runs on the latest react and styled-components (v4), so it should be much faster.
Moreover, you can start using the newest React's features: hooks, memo, suspense, etc. I suggest to use them while creating new components.

#### What we didn't do

- material-ui upgrade - a lot of breaking changes. It'll make this PR even bigger.
- Binding CRA to handle the whole stack - It's a good idea, but we'll have to handle some custom solution used in your project, which will take some time.
- Merging global viewer with viewer service - let me know if you even want to do that. It's doable, but we tried to leave your globals as they are.

#### Known issues
The new version of styled-components sometimes throws warnings (development mode only) for custom event handlers e.g `onRootClick`. We checked that is the issue on the library side, but the good news is the next version is coming. It'll be faster (they rewrote styled-components using hooks) and contains a bugfix that we need. ;)

We went through for almost every place in the app during this part of the migration, so you probably will need to do the same while testing.

### CODE REVIEW

**Comment by simplypixi** [comment]
@carmenfan @sanmont3drepo @Charence This PR is enormous - we know that, but we couldn't split that to smaller chunks, because these functionalities were related.
The only idea I have is moving the latest stack changes to separate branch - I ran auto-fix after re-configuring the linter, so a lot of files were updated.

Let me know what you think.

**Comment by simplypixi** [comment]
> On some models I see the following errors:
> 
> Screenshot 2019-07-17 12 17 32
> 
> The above appears when I try to view the model. The other problem is the error message shouldn't display the tags such as `<br>`.
> 
> Screenshot 2019-07-17 12 19 50
> Screenshot 2019-07-17 12 19 05
> 
> The above appears when I try to proceed and click on an issue.

Could you check if it occurs on tree migration branch? Is it possible the model doesn't have a name? Because it looks like that (I mean about the error on the dialog). :P

**Comment by dleszcz** [comment]
> On some models I see the following errors:
> 
> Screenshot 2019-07-17 12 17 32
> 
> The above appears when I try to view the model. The other problem is the error message shouldn't display the tags such as `<br>`.
> 
> Screenshot 2019-07-17 12 19 50
> Screenshot 2019-07-17 12 19 05
> 
> The above appears when I try to proceed and click on an issue.

@Charence These errors are fixed on Tree pull request in the last commit

**Comment by simplypixi** [comment]
> The above appears when I try to view the model. The other problem is the error message shouldn't display the tags such as `<br>`.

And this one works too. It looks like that:
image

Did you test it on latest changes?

**Comment by Charence** [comment]
> Did you test it on latest changes?

@simplypixi @dleszcz Which branch did you apply the changes to? ISSUE_1590?

**Comment by simplypixi** [comment]
> > Did you test it on latest changes?
> 
> @simplypixi @dleszcz Which branch did you apply the changes to? ISSUE_1590?

Yes, `ISSUE_1590`.

I switched the `base branch` to ISSUE_1475 (Tree branch). ISSUE_1590 is a part of main feature branch but since we did rebase it's not necessary to merge it to the parent. I mean it contains all changes.

We'll resolve the conflicts as soon as possible.

File: frontend/globals/viewer.ts [hunk 1]
```diff
@@ -287,6 +267,22 @@ export class Viewer {
...
+		if (progress === 1) {
+			this.emit(Viewer.EVENT.VIEWER_INIT_SUCCESS, progress);
+		} else {
+			this.emit(Viewer.EVENT.VIEWER_INIT_PROGRESS, progress);
+		}
+	}
+
+	public onModelProgress = (progress) => {
+		if (progress === 1) {
+			this.emit(Viewer.EVENT.MODEL_LOADED, progress);
```
Comment (carmenfan) [inline frontend/globals/viewer.ts]: I don't think this should be emitted here. We should stick to when unity actually sends you the model loaded signal.

File: frontend/index.html [hunk 2]
```diff
@@ -98,40 +120,7 @@ <h4 style="font-weight: lighter;">
...
-      }
-
-    </script>
-
-
-
-    <!-- Entry point for all components-->
-    <div ui-view="main" flex layout="column"></div>
-    <app />
+    <div id="app"></div>
```
Comment (carmenfan) [inline frontend/index.html]: Why is this removed? I can no longer see the internet explorer message (right is staging):
image

File: frontend/internals/webpack/tools/loaders.js [hunk 3]
```diff
@@ -105,5 +117,6 @@ module.exports = {
   FontLoader,
   ImageLoader,
   HTMLLoader,
-  PugLoader
+  PugLoader,
```
Comment (carmenfan) [inline frontend/internals/webpack/tools/loaders.js]: Do we still need PugLoader?

**Comment by sanmont3drepo** [comment]
**Visual Settings**
- [ ] After reloading from changing visual settings the visual settings are not applied
- [ ] After changing the visual settings, saving them, then clicking the reset and saving, the app gets blank:
image

File: frontend/modules/dialog/dialog.redux.ts [hunk 4]
```diff
@@ -104,20 +113,37 @@ export const showScreenshotDialog = (state = INITIAL_STATE, action) => {
...
+const hideDialog = (state = INITIAL_STATE) => {
 	return { ...state, isOpen: false, isPending: false };
 };
 
-export const setPendingState = (state = INITIAL_STATE, {isPending}) => {
+const setPendingState = (state = INITIAL_STATE, {isPending}) => {
 	return { ...state, isPending };
 };
 
+const setMuteNotifications = (state = INITIAL_STATE, { muteNotifications }) => {
```
Comment (carmenfan) [inline frontend/modules/dialog/dialog.redux.ts]: I think we got rid of the dialog that uses this? Is it still needed?

**Comment by simplypixi** [comment]
@carmenfan @dleszcz merged Tree branch yesterday (to resolve conflicts) and he had a problem with starting BE, so he pushed merged changes as a new branch (ISSUE_1590_copy). I checked that today and I have the same issue (required port is free). Did you change something recently? Could you look at it?
image

File: frontend/globals/viewer.ts (hunk 1 above)
Comment (simplypixi) [reply frontend/globals/viewer.ts]: fixed

File: frontend/index.html (hunk 2 above)
Comment (simplypixi) [reply frontend/index.html]: We did it in a "better way." Whole browser detection/unsupported page logic was moved to `support.ts` (it's not included directly in the template), but maybe we set wrong IE or Edge versions. Which versions of IE should we mark as unsupported? Do you support Edge?

image

File: frontend/internals/webpack/tools/loaders.js (hunk 3 above)
Comment (simplypixi) [reply frontend/internals/webpack/tools/loaders.js]: I assume no

Comment (simplypixi) [reply frontend/internals/webpack/tools/loaders.js]: removed

File: frontend/modules/dialog/dialog.redux.ts (hunk 4 above)
Comment (simplypixi) [reply frontend/modules/dialog/dialog.redux.ts]: removed

**Comment by simplypixi** [comment]
> * There seems to be an async problem. Sometimes if you hit refresh on a model(more likely on a hard refresh), you get this:

@carmenfan is it happening on certain models or all? I can't reproduce this bug :(

File: frontend/index.html (hunk 2 above)
Comment (carmenfan) [reply frontend/index.html]: We dont support any version of ie (they generally don't support webgl),  but we need to make sure whatever code you use for this msg it is visible for anything as low as ie6. we don't support edge either but it should work (ie they can use it but if they run into bugs we dont fix them..) , I can't remember what we currently do with it. I think we just don't say anyhthing for edge and let the user through.

**Comment by carmenfan** [comment]
@simplypixi the port problem is the problem I mentioned in the email. You need to update your chat config to define the port as 'port' instead of 'http_port' or 'https_port'

**Comment by simplypixi** [comment]
> @simplypixi the port problem is the problem I mentioned in the email. You need to update your chat config to define the port as 'port' instead of 'http_port' or 'https_port'

Makes sense now. I totally forgot about that the Tree branch contains staging changes already.
Thanks!

File: frontend/index.html (hunk 2 above)
Comment (simplypixi) [reply frontend/index.html]: @carmenfan Unsupported browser screen shouldn't be displayed for each version of IE now. Could you check it? If it still doesn't work properly, let me know which version od IE and Windows you use.

**Comment by simplypixi** [comment]
> > * There seems to be an async problem. Sometimes if you hit refresh on a model(more likely on a hard refresh), you get this:
> 
> @carmenfan is it happening on certain models or all? I can't reproduce this bug :(

@carmenfan I'm still not able to reproduce this bug, but I've changed a bit logic of loading unityLoader script. Could you check if it's happening?

**Comment by simplypixi** [comment]
@carmenfan Uff, I resolved conflicts. Please double-check the "resources" feature. Moreover, two unity changes were conflicted (8a814ed76, 1119f19bd) so probably you will have to prepare a new build.

@sanmont3drepo Could you migrate your change from staging? We discussed it before on Tree PR. If not, let us know. ;)

```
this.$scope.$watch(() => this.colorOverrides,  async (overrides, prevOverrides) => {
	const toAdd = overridesDiff(overrides, prevOverrides);
	const toRemove = overridesDiff(prevOverrides, overrides );

	removeColorOverrides(toRemove);
	addColorOverrides(toAdd);
});
```

**Comment by carmenfan** [comment]
> @carmenfan I'm still not able to reproduce this bug, but I've changed a bit logic of loading unityLoader script. Could you check if it's happening?

I just tried it, and it seems like it's no longer happening!

> @carmenfan Uff, I resolved conflicts. Please double-check the "resources" feature. Moreover, two unity changes were conflicted (8a814ed, 1119f19) so probably you will have to prepare a new build.

OK.. I'm working on my unity chanage at the moment, depending on the timing of how everything is merged, you might find more conlicts... 😸 

resources seems to be still working 👍 

> @sanmont3drepo Could you migrate your change from staging? We discussed it before on Tree PR. If not, let us know. ;)

@sanmont3drepo  is on holiday until next monday, so he won't be able to do anything until then.

**Comment by carmenfan** [comment]
@simplypixi is there anything outstanding on this?  It looks ok to me, perhaps we should merge this to tree migration?

**Comment by simplypixi** [comment]
> @sanmont3drepo is on holiday until next monday, so he won't be able to do anything until then.

No problem, I can migrate that. 👍 

> @simplypixi is there anything outstanding on this? It looks ok to me, perhaps we should merge this to tree migration?

Tbh, it's up to you. We can merge it to tree migration :)
Did you go through the whole app? I mean, we changed "app core", so it'll be good to do some smoke tests at least.

The most important places to test: Tree panel, Viewer toolbar, app routing
The rest should be fine.

**Comment by carmenfan** [comment]
> > @sanmont3drepo is on holiday until next monday, so he won't be able to do anything until then.
> 
> No problem, I can migrate that. 👍
> 
> > @simplypixi is there anything outstanding on this? It looks ok to me, perhaps we should merge this to tree migration?
> 
> Tbh, it's up to you. We can merge it to tree migration :)
> Did you go through the whole app? I mean, we changed "app core", so it'll be good to do some smoke tests at least.
> 
> The most important places to test: Tree panel, Viewer toolbar, app routing
> The rest should be fine.

I did a quick test and didn't really see anything wrong with the things you mentioned, I'd say lets merge this into the tree and give the tree more testing.

Otherwise i feel like we're testing a lot of versions of everything that may once again need more testing when it's merged together.. 😆

**Comment by simplypixi** [comment]
> Otherwise i feel like we're testing a lot of versions of everything that may once again need more testing when it's merged together.. 😆
Hehe, I understand.... :D

**Pull request merged**